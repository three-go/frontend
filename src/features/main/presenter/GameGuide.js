import React from "react";

import { StyleSheet, View, Text, ScrollView, Image } from "react-native";
import Icon from "react-native-vector-icons/MaterialCommunityIcons";
import PropTypes from "prop-types";

import { colors } from "../../../common/constants";

const GameGuide = ({ item, style }) => {
  return (
    <View style={[styles.container, style]}>
      <ScrollView showsVerticalScrollIndicator={false}>
        <View>
          <Text style={styles.title}>{item.title}</Text>
          <Text style={styles.content}>{item.titleContent}</Text>
        </View>
        {item.children.map((value) => {
          return (
            <View key={value.id}>
              <View style={styles.imageContainer}>
                <Text style={styles.title}>{value.title}</Text>
                <Image style={styles.image} source={value.image} />
                {value.icon && (
                  <Icon style={styles.icon} name={value.icon} size={100} />
                )}
              </View>
              <View style={styles.contentContainer}>
                <Text style={styles.content}>{value.content}</Text>
              </View>
            </View>
          );
        })}
      </ScrollView>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    justifyContent: "center",
    alignItems: "center",
    width: "100%",
    padding: 5,
    backgroundColor: colors.dark,
  },
  title: {
    marginTop: 10,
    marginBottom: 10,
    fontSize: 15,
    fontWeight: "900",
    color: colors.ivory,
  },
  content: {
    marginBottom: 10,
    fontSize: 12,
    color: colors.ivory,
  },
  imageContainer: {
    flex: 1,
    alignItems: "center",
    width: "100%",
    height: 500,
    borderTopWidth: 5,
    borderTopColor: colors.ivory,
  },
  image: {
    flex: 1,
    width: "100%",
    height: "100%",
    resizeMode: "contain",
  },
  contentContainer: {
    flex: 1,
  },
  icon: {
    position: "absolute",
    top: 200,
    left: 63,
  },
});

export default GameGuide;

GameGuide.propTypes = {
  item: PropTypes.shape({
    children: PropTypes.arrayOf(
      PropTypes.shape({
        content: PropTypes.string.isRequired,
        icon: PropTypes.string,
        id: PropTypes.string.isRequired,
        image: PropTypes.number.isRequired,
        title: PropTypes.string.isRequired,
      })
    ),
    title: PropTypes.string.isRequired,
    titleContent: PropTypes.string.isRequired,
  }),
  style: PropTypes.object,
};
