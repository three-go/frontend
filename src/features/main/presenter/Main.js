import React from "react";

import { View, StyleSheet, StatusBar } from "react-native";
import { SafeAreaView } from "react-native-safe-area-context";

import { game } from "../../../common";
import { DefaultButton, SquareButton, Logo } from "../../../components";

const Main = ({
  onExitApp,
  onShowScoreModal,
  onSetGameNameAndShowDescriptionModal,
}) => {
  return (
    <SafeAreaView style={styles.container}>
      <StatusBar backgroundColor="#212529" barStyle="light-content" />

      <View>
        <Logo style={styles.logo} size="lg" />
      </View>

      <View style={styles.gameWrapper}>
        {game.keys.map((v) => {
          return (
            <SquareButton
              key={v}
              content={game.names[v]}
              color={game.colors[v]}
              onPress={onSetGameNameAndShowDescriptionModal(v)}
            />
          );
        })}
      </View>

      <View style={styles.button}>
        <DefaultButton
          content="점수 보기"
          color="#96A1A8"
          onPress={onShowScoreModal}
        />
      </View>

      <View style={styles.button}>
        <DefaultButton
          content="게임 종료"
          color="#c92a2a"
          onPress={onExitApp}
        />
      </View>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    backgroundColor: "#212529",
  },
  statusBar: {
    backgroundColor: "#212529",
  },
  gameWrapper: {
    flexDirection: "row",
    justifyContent: "space-evenly",
    width: "100%",
    height: 100,
    marginTop: 50,
    marginBottom: 100,
  },
  button: {
    marginBottom: 20,
  },
});

export default Main;
