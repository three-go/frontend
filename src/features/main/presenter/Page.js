import React from "react";

import { StyleSheet, View, Text, ScrollView } from "react-native";

const Page = ({ item, style }) => {
  return (
    <View style={[styles.container, style]}>
      <ScrollView showsVerticalScrollIndicator={false}>
        {item.length ? (
          item.map((value, index) => {
            return (
              <View key={value.id} style={styles.wrapper(index)}>
                <Text style={styles.text}>{value.name}</Text>
                <Text style={styles.text}>{value.score}</Text>
              </View>
            );
          })
        ) : (
          <View style={styles.empty}>
            <Text style={styles.text}>등록된 정보가 없습니다.</Text>
          </View>
        )}
      </ScrollView>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    justifyContent: "center",
    alignItems: "center",
    width: "100%",
    backgroundColor: "#FCF8F6",
  },
  wrapper: (index) => {
    return {
      justifyContent: "center",
      alignItems: "center",
      flexDirection: "row",
      width: "100%",
      height: 50,
      borderWidth: 2,
      borderRadius: 5,
      borderColor: "#FCF8F6",
      backgroundColor: index % 2 === 0 ? "#96A1A8" : "#FCF8F6",
    };
  },
  text: {
    width: "50%",
    color: "#212529",
    textAlign: "center",
    fontSize: 15,
    fontWeight: "bold",
  },
  empty: {
    flexDirection: "row",
    justifyContent: "center",
    alignItems: "center",
    width: "100%",
    height: 50,
    minHeight: "100%",
  },
});

export default Page;
