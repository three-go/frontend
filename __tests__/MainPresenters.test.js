import React from "react";

import { fireEvent, render } from "@testing-library/react-native";
import "@testing-library/jest-native";
import "react-native-gesture-handler/jestSetup";
import RNExitApp from "react-native-exit-app";

import GameGuide from "../src/features/main/presenter/GameGuide";
import Main from "../src/features/main/presenter/Main";
import Page from "../src/features/main/presenter/Page";

jest.mock("react-native-sound", () => {
  const Sound = () => {
    const obj = {};
    obj.sound = "music";
    obj.setNumberOfLoops = jest.fn();
    obj.play = jest.fn();
    return obj;
  };

  return Sound;
});

const size = {
  SCREEN_WIDTH: 300,
  GAP: 15,
  OFFSET: 15,
  PAGE_WIDTH: 300 - (15 + 15) * 2,
};

describe("Presenter Test", () => {
  it("Main Test", () => {
    const mockExitApp = jest.spyOn(RNExitApp, "exitApp");
    mockExitApp.mockReturnValue("exit");
    const mockShowScoreModal = jest.fn();
    const mockSetGameNameAndShowDescriptionModal = jest.fn();

    const props = {
      onExitApp: () => {
        RNExitApp.exitApp();
      },
      onShowScoreModal: mockShowScoreModal,
      onSetGameNameAndShowDescriptionModal:
        mockSetGameNameAndShowDescriptionModal,
    };

    const main = render(<Main {...props} />);
    const mainJson = main.toJSON();
    const faceGoButton = main.getByText("Face Go");
    const shoutGoButton = main.getByText("Shout Go");
    const scoreButton = main.getByText("점수 보기");
    const exitButton = main.getByText("게임 종료");

    expect(faceGoButton).toHaveTextContent("Face Go");
    expect(shoutGoButton).toHaveTextContent("Shout Go");
    expect(scoreButton).toHaveTextContent("점수 보기");
    expect(exitButton).toHaveTextContent("게임 종료");
    expect(mainJson).toBeTruthy();
    expect(mainJson).toMatchSnapshot();

    fireEvent.press(exitButton);
    expect(mockExitApp).toHaveBeenCalled();

    fireEvent.press(scoreButton);
    expect(mockShowScoreModal).toHaveBeenCalled();

    fireEvent.press(faceGoButton);
    expect(mockSetGameNameAndShowDescriptionModal).toHaveBeenCalled();

    fireEvent.press(shoutGoButton);
    expect(mockSetGameNameAndShowDescriptionModal).toHaveBeenCalled();
  });

  it("GameGuide Test", () => {
    const props = {
      item: {
        title: "게임설명",
        titleContent: "게임방법",
        children: [
          {
            id: "id1",
            title: "조작",
            content: "조작설명",
            image: 0,
          },
        ],
      },
      style: {
        papeWrapper: (prop) => {
          return {
            width: prop.PAGE_WIDTH,
            marginHorizontal: prop.GAP / 2,
          };
        },
      },
    };
    const gameGuide = render(
      <GameGuide item={props.item} style={props.style.papeWrapper(size)} />
    );
    const gameGuideJson = gameGuide.toJSON();

    expect(gameGuide.getByText("게임설명")).toHaveTextContent("게임설명");
    expect(gameGuide.getByText("게임방법")).toHaveTextContent("게임방법");
    expect(gameGuide.getByText("조작")).toHaveTextContent("조작");
    expect(gameGuide.getByText("조작설명")).toHaveTextContent("조작설명");
    expect(gameGuideJson).toBeTruthy();
    expect(gameGuideJson).toMatchSnapshot();
  });

  it("Page Test", () => {
    const props = {
      item: [
        { id: "id1", name: "John", score: 300 },
        { id: "id2", name: "Tom", score: 500 },
      ],
      style: {
        pageWrapper: (prop) => {
          return {
            width: prop.PAGE_WIDTH,
            marginHorizontal: prop.GAP / 2,
          };
        },
      },
    };
    const page = render(<Page {...props} />);
    const pageJson = page.toJSON();
    const score1 = page.getByText("300");
    const score2 = page.getByText("500");
    const name1 = page.getByText("John");
    const name2 = page.getByText("Tom");

    expect(score1).toHaveTextContent("300");
    expect(score2).toHaveTextContent("500");
    expect(name1).toHaveTextContent("John");
    expect(name2).toHaveTextContent("Tom");
    expect(pageJson).toBeTruthy();
    expect(pageJson).toMatchSnapshot();

    const emptyProps = {
      item: [],
      style: {
        pageWrapper: (prop) => {
          return {
            width: prop.PAGE_WIDTH,
            marginHorizontal: prop.GAP / 2,
          };
        },
      },
    };
    const emptyPage = render(<Page {...emptyProps} />);
    const emptyPageJson = emptyPage.toJSON();
    const emptyPageText = emptyPage.getByText("등록된 정보가 없습니다.");

    expect(emptyPageText).toHaveTextContent("등록된 정보가 없습니다.");
    expect(emptyPageJson).toBeTruthy();
    expect(emptyPageJson).toMatchSnapshot();
  });
});
