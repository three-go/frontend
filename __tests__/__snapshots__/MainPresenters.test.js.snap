// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Presenter Test GameGuide Test 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#212529",
        "justifyContent": "center",
        "padding": 5,
        "width": "100%",
      },
      Object {
        "marginHorizontal": 7.5,
        "width": 240,
      },
    ]
  }
>
  <RCTScrollView
    showsVerticalScrollIndicator={false}
  >
    <View>
      <View>
        <Text
          style={
            Object {
              "color": "#FCF8F6",
              "fontSize": 15,
              "fontWeight": "900",
              "marginBottom": 10,
              "marginTop": 10,
            }
          }
        >
          게임설명
        </Text>
        <Text
          style={
            Object {
              "color": "#FCF8F6",
              "fontSize": 12,
              "marginBottom": 10,
            }
          }
        >
          게임방법
        </Text>
      </View>
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "borderTopColor": "#FCF8F6",
              "borderTopWidth": 5,
              "flex": 1,
              "height": 500,
              "width": "100%",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#FCF8F6",
                "fontSize": 15,
                "fontWeight": "900",
                "marginBottom": 10,
                "marginTop": 10,
              }
            }
          >
            조작
          </Text>
          <Image
            source={0}
            style={
              Object {
                "flex": 1,
                "height": "100%",
                "resizeMode": "contain",
                "width": "100%",
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#FCF8F6",
                "fontSize": 12,
                "marginBottom": 10,
              }
            }
          >
            조작설명
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Presenter Test Main Test 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "bottom",
      "left",
      "right",
      "top",
    ]
  }
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#212529",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View>
    <Image
      resizeMode="contain"
      source={
        Object {
          "testUri": "../../../public/assets/logo.png",
        }
      }
      style={
        Object {
          "height": 160,
          "width": 240,
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "height": 100,
        "justifyContent": "space-evenly",
        "marginBottom": 100,
        "marginTop": 50,
        "width": "100%",
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#00BBD1",
          "borderColor": "#FCF8F6",
          "borderRadius": 4,
          "borderWidth": 4,
          "height": 120,
          "justifyContent": "center",
          "opacity": 1,
          "width": 120,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#FCF8F6",
            "fontSize": 18,
            "fontWeight": "bold",
            "padding": 8,
          }
        }
      >
        Face Go
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#00E0EA",
          "borderColor": "#FCF8F6",
          "borderRadius": 4,
          "borderWidth": 4,
          "height": 120,
          "justifyContent": "center",
          "opacity": 1,
          "width": 120,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#FCF8F6",
            "fontSize": 18,
            "fontWeight": "bold",
            "padding": 8,
          }
        }
      >
        Shout Go
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "marginBottom": 20,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#96A1A8",
          "borderRadius": 4,
          "height": 40,
          "justifyContent": "center",
          "opacity": 1,
          "width": 150,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#FCF8F6",
            "fontSize": 15,
            "fontWeight": "bold",
            "padding": 8,
          }
        }
      >
        점수 보기
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "marginBottom": 20,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#f45a5a",
          "borderRadius": 4,
          "height": 40,
          "justifyContent": "center",
          "opacity": 1,
          "width": 150,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#FCF8F6",
            "fontSize": 15,
            "fontWeight": "bold",
            "padding": 8,
          }
        }
      >
        게임 종료
      </Text>
    </View>
  </View>
</RNCSafeAreaView>
`;

exports[`Presenter Test Page Test 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#FCF8F6",
        "justifyContent": "center",
        "width": "100%",
      },
      Object {
        "pageWrapper": [Function],
      },
    ]
  }
>
  <RCTScrollView
    showsVerticalScrollIndicator={false}
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#96A1A8",
            "borderColor": "#FCF8F6",
            "borderRadius": 5,
            "borderWidth": 2,
            "flexDirection": "row",
            "height": 50,
            "justifyContent": "center",
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#212529",
              "fontSize": 15,
              "fontWeight": "bold",
              "textAlign": "center",
              "width": "50%",
            }
          }
        >
          John
        </Text>
        <Text
          style={
            Object {
              "color": "#212529",
              "fontSize": 15,
              "fontWeight": "bold",
              "textAlign": "center",
              "width": "50%",
            }
          }
        >
          300
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#FCF8F6",
            "borderColor": "#FCF8F6",
            "borderRadius": 5,
            "borderWidth": 2,
            "flexDirection": "row",
            "height": 50,
            "justifyContent": "center",
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#212529",
              "fontSize": 15,
              "fontWeight": "bold",
              "textAlign": "center",
              "width": "50%",
            }
          }
        >
          Tom
        </Text>
        <Text
          style={
            Object {
              "color": "#212529",
              "fontSize": 15,
              "fontWeight": "bold",
              "textAlign": "center",
              "width": "50%",
            }
          }
        >
          500
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Presenter Test Page Test 2`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#FCF8F6",
        "justifyContent": "center",
        "width": "100%",
      },
      Object {
        "pageWrapper": [Function],
      },
    ]
  }
>
  <RCTScrollView
    showsVerticalScrollIndicator={false}
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 50,
            "justifyContent": "center",
            "minHeight": "100%",
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#212529",
              "fontSize": 15,
              "fontWeight": "bold",
              "textAlign": "center",
              "width": "50%",
            }
          }
        >
          등록된 정보가 없습니다.
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
